File: DS/Graph.py
import sys
import collections
import heapq

# --- CP TEMPLATE START ---

# For faster I/O
# sys.stdin = open('in.txt', 'r')
# sys.stdout = open('out.txt', 'w')

# It's recommended to increase recursion limit for deep DFS problems
# sys.setrecursionlimit(10**6)

# Adjacency list representation
# For unweighted graphs: adj = collections.defaultdict(list)
# For weighted graphs: adj = collections.defaultdict(list) where adj[u].append((v, w))

INF = float('inf')

# --- Disjoint Set Union (DSU) / Union-Find ---
class DSU:
    def __init__(self, n):
        # Using 1-based indexing for n nodes
        self.parent = list(range(n + 1))

    def find(self, i):
        if self.parent[i] == i:
            return i
        # Path compression
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def unite(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j

# --- Breadth-First Search (BFS) ---
# Use for shortest path in unweighted graphs.
# Returns a list of distances from the start node.
def bfs(start_node, n, adj):
    # Using 1-based indexing
    dist = [-1] * (n + 1)
    q = collections.deque()

    dist[start_node] = 0
    q.append(start_node)

    while q:
        u = q.popleft()

        for v in adj[u]:
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                q.append(v)
    return dist

# --- Depth-First Search (DFS) ---
# Use for connectivity, cycle detection, path finding, etc.
def dfs_util(u, visited, adj):
    visited[u] = True
    # Process node u here
    # print(f"Visiting node {u}")
    for v in adj[u]:
        if not visited[v]:
            dfs_util(v, visited, adj)

def dfs(start_node, n, adj):
    # Using 1-based indexing
    visited = [False] * (n + 1)
    dfs_util(start_node, visited, adj)

# --- Dijkstra's Algorithm ---
# Use for shortest path in weighted graphs with non-negative weights.
# Returns a list of shortest distances from the start node.
def dijkstra(start_node, n, adj):
    # Using 1-based indexing
    dist = [INF] * (n + 1)
    dist[start_node] = 0
    
    # Min-priority queue storing (distance, node)
    pq = [(0, start_node)]

    while pq:
        d, u = heapq.heappop(pq)

        if d > dist[u]:
            continue

        for v, weight in adj[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                heapq.heappush(pq, (dist[v], v))
    return dist

# --- TEMPLATE END ---

def main():
    """Example Usage"""
    n = 5  # Number of nodes

    # --- DSU Example ---
    print("DSU Example:")
    dsu = DSU(n)
    dsu.unite(1, 2)
    dsu.unite(3, 4)
    print(f"Find(1): {dsu.find(1)}, Find(2): {dsu.find(2)}")
    print(f"Find(3): {dsu.find(3)}, Find(4): {dsu.find(4)}")
    print(f"Are 1 and 4 connected? {'Yes' if dsu.find(1) == dsu.find(4) else 'No'}")
    dsu.unite(1, 4)
    print(f"Are 1 and 4 connected now? {'Yes' if dsu.find(1) == dsu.find(4) else 'No'}")
    print("---")

    # --- BFS/DFS Example ---
    print("BFS/DFS Example:")
    adj_unweighted = collections.defaultdict(list)
    edges = [(1, 2), (2, 3), (3, 4), (4, 5)]
    for u, v in edges:
        adj_unweighted[u].append(v)
        adj_unweighted[v].append(u)

    bfs_dist = bfs(1, n, adj_unweighted)
    print(f"Shortest distance from 1 to 5 (BFS): {bfs_dist[5]}")
    print("---")

    # --- Dijkstra Example ---
    print("Dijkstra Example:")
    adj_weighted = collections.defaultdict(list)
    weighted_edges = [
        (1, 2, 10), (1, 3, 3), (2, 3, 1), (2, 4, 2),
        (3, 2, 4), (3, 4, 8), (3, 5, 2), (4, 5, 7), (5, 4, 9)
    ]
    for u, v, w in weighted_edges:
        adj_weighted[u].append((v, w))

    dijkstra_dist = dijkstra(1, n, adj_weighted)
    print(f"Shortest distance from 1 to 5 (Dijkstra): {dijkstra_dist[5]}") # Expected: 3+2=5

if __name__ == "__main__":
    main()

File: DS/Graph.cpp
#include<bits/stdc++.h>
#define ll long long
using namespace std;
// --- CP TEMPLATE START ---

// Common type aliases

using pii = std::pair<int, int>;
using pll = std::pair<ll, ll>;

// Adjacency list representation
// For unweighted graphs: std::vector<std::vector<int>> adj;
// For weighted graphs: std::vector<std::vector<pii>> adj; (stores {neighbor, weight})

const int INF_INT = 1e9;
const ll INF_LL = 1e18;

// --- Disjoint Set Union (DSU) / Union-Find ---
struct DSU {
    std::vector<int> parent;
    DSU(int n) {
        parent.resize(n + 1);
        std::iota(parent.begin(), parent.end(), 0); // Fills with 0, 1, 2, ...
    }
    int find(int i) {
        if (parent[i] == i)
            return i;
        return parent[i] = find(parent[i]); // Path compression
    }
    void unite(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);
        if (root_i != root_j) {
            parent[root_i] = root_j;
        }
    }
};

// --- Breadth-First Search (BFS) ---
// Use for shortest path in unweighted graphs.
// Returns a vector of distances from the start node.
std::vector<int> bfs(int start_node, int n, const std::vector<std::vector<int>>& adj) {
    std::vector<int> dist(n + 1, -1);
    std::queue<int> q;

    dist[start_node] = 0;
    q.push(start_node);

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v : adj[u]) {
            if (dist[v] == -1) {
                dist[v] = dist[u] + 1;
                q.push(v);
            }
        }
    }
    return dist;
}

// --- Depth-First Search (DFS) ---
// Use for connectivity, cycle detection, path finding, etc.
void dfs_util(int u, std::vector<bool>& visited, const std::vector<std::vector<int>>& adj) {
    visited[u] = true;
    // Process node u here
    for (int v : adj[u]) {
        if (!visited[v]) {
            dfs_util(v, visited, adj);
        }
    }
}

void dfs(int start_node, int n, const std::vector<std::vector<int>>& adj) {
    std::vector<bool> visited(n + 1, false);
    dfs_util(start_node, visited, adj);
}

// --- Dijkstra's Algorithm ---
// Use for shortest path in weighted graphs with non-negative weights.
// Returns a vector of shortest distances from the start node.
std::vector<ll> dijkstra(int start_node, int n, const std::vector<std::vector<pll>>& adj) {
    std::vector<ll> dist(n + 1, INF_LL);
    dist[start_node] = 0;
    
    // Min-priority queue storing {distance, node}
    std::priority_queue<pll, std::vector<pll>, std::greater<pll>> pq;
    pq.push({0, start_node});

    while (!pq.empty()) {
        ll d = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        if (d > dist[u]) {
            continue;
        }

        for (auto& edge : adj[u]) {
            int v = edge.first;
            ll weight = edge.second;
            if (dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

// --- TEMPLATE END ---

void solve() {
    // Example Usage
    int n = 5; // Number of nodes
    
    // --- DSU Example ---
    std::cout << "DSU Example:" << std::endl;
    DSU dsu(n);
    dsu.unite(1, 2);
    dsu.unite(3, 4);
    std::cout << "Find(1): " << dsu.find(1) << ", Find(2): " << dsu.find(2) << std::endl;
    std::cout << "Find(3): " << dsu.find(3) << ", Find(4): " << dsu.find(4) << std::endl;
    std::cout << "Are 1 and 4 connected? " << (dsu.find(1) == dsu.find(4) ? "Yes" : "No") << std::endl;
    dsu.unite(1, 4);
    std::cout << "Are 1 and 4 connected now? " << (dsu.find(1) == dsu.find(4) ? "Yes" : "No") << std::endl;
    std::cout << "---" << std::endl;

    // --- BFS/DFS Example ---
    std::cout << "BFS/DFS Example:" << std::endl;
    std::vector<std::vector<int>> adj_unweighted(n + 1);
    adj_unweighted[1].push_back(2);
    adj_unweighted[2].push_back(1);
    adj_unweighted[2].push_back(3);
    adj_unweighted[3].push_back(2);
    adj_unweighted[3].push_back(4);
    adj_unweighted[4].push_back(3);
    adj_unweighted[4].push_back(5);
    adj_unweighted[5].push_back(4);

    std::vector<int> bfs_dist = bfs(1, n, adj_unweighted);
    std::cout << "Shortest distance from 1 to 5 (BFS): " << bfs_dist[5] << std::endl;
    std::cout << "---" << std::endl;

    // --- Dijkstra Example ---
    std::cout << "Dijkstra Example:" << std::endl;
    std::vector<std::vector<pll>> adj_weighted(n + 1);
    adj_weighted[1].push_back({2, 10});
    adj_weighted[1].push_back({3, 3});
    adj_weighted[2].push_back({3, 1});
    adj_weighted[2].push_back({4, 2});
    adj_weighted[3].push_back({2, 4});
    adj_weighted[3].push_back({4, 8});
    adj_weighted[3].push_back({5, 2});
    adj_weighted[4].push_back({5, 7});
    adj_weighted[5].push_back({4, 9});

    std::vector<ll> dijkstra_dist = dijkstra(1, n, adj_weighted);
    std::cout << "Shortest distance from 1 to 5 (Dijkstra): " << dijkstra_dist[5] << std::endl; // Expected: 3+2=5
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    solve();
    return 0;
}
